x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/RemoteAccessAuthenticationService.java

```diff
@@ -0,0 +1,228 @@
+        ) {
+            final Supplier<ThreadContext.StoredContext> storedContextSupplier = threadContext.newRestorableContext(false);
+            authcContext.addAuthenticationToken(remoteAccessHeaders.clusterCredential());
+            authenticationService.authenticate(
+                authcContext,
+                new ContextPreservingActionListener<>(storedContextSupplier, ActionListener.wrap(authentication -> {
+                    assert authentication.isApiKey() : "initial authentication for remote access must be by API key";
+                    final RemoteAccessAuthentication remoteAccessAuthentication = remoteAccessHeaders.remoteAccessAuthentication();
+                    final Subject receivedEffectiveSubject = remoteAccessAuthentication.getAuthentication().getEffectiveSubject();
+                    final User user = receivedEffectiveSubject.getUser();
+
+                    final Authentication finalAuthentication;
+                    if (SystemUser.is(user)) {
+                        finalAuthentication = authentication.toRemoteAccess(
+                            new RemoteAccessAuthentication(
+                                Authentication.newInternalAuthentication(
+                                    SystemUser.INSTANCE,
+                                    receivedEffectiveSubject.getTransportVersion(),
+                                    receivedEffectiveSubject.getRealm().getNodeName()
+                                ),
+                                new RoleDescriptorsIntersection(CROSS_CLUSTER_INTERNAL_ROLE)
+                            )
+                        );
+                    } else if (User.isInternal(user)) {
+                        throw new IllegalArgumentException(
+                            "received cross cluster request from an unexpected internal user [" + user.principal() + "]"
+                        );
+                    } else {
+                        finalAuthentication = authentication.toRemoteAccess(remoteAccessAuthentication);
+                    }
+
+                    writeAuthToContext(authcContext, finalAuthentication, listener);
+                }, ex -> withRequestProcessingFailure(authcContext, ex, listener)))
+            );
+        }
+    }
+
+    public AuthenticationService getAuthenticationService() {
+        return authenticationService;
+    }
+
+    private record RemoteAccessHeaders(
+        ApiKeyService.ApiKeyCredentials clusterCredential,
+        RemoteAccessAuthentication remoteAccessAuthentication
+    ) {
+        private void validate() {
+            final String subjectPrincipal = remoteAccessAuthentication.getAuthentication().getEffectiveSubject().getUser().principal();
+            for (RemoteAccessAuthentication.RoleDescriptorsBytes roleDescriptorsBytes : remoteAccessAuthentication
+                .getRoleDescriptorsBytesList()) {
+                final Set<RoleDescriptor> roleDescriptors = roleDescriptorsBytes.toRoleDescriptors();
```

Comment:
We can for sure revisit this, but the reason I validate here is because it gives us an early place to fail and because the other sensible place to validate (where we can avoid the extra deserialization) is all the way in [resolveRemoteAccessRoleReference](https://github.com/elastic/elasticsearch/blob/c1b0bf65cfc383872823ea87dd8b3ce0ac098cc2/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/RoleDescriptorStore.java#L136). The problem with that is that at that stage, it's actually valid for a role descriptor to have a cluster privilege (currently `monitor`, because of the internal `_cross_cluster_internal` role) and working around that cleanly does not seem worth prioritization. Let me know if you feel strongly, but I'd prefer to take the performance hit for now and revisit as a possible optimization in the future.

Reply:
hmm, should we just send indices privileges so this is guaranteed by deserialization itself? I under this would be quite a big change. So really just a talking point for now.

A more practical approach is to _not_ use `RemoteAccessRoleReference` if the remote user is `_system`. We can have a different role reference type, e.g. `StaticRoleReference`, that directly gives out the roleDescriptors without going through the de/serialization detour at all. This is more efficient and should solve the dilemma here.

Reply:
> hmm, should we just send indices privileges so this is guaranteed by deserialization itself? I under this would be quite a big change. So really just a talking point for now.

This would be possible, but as you mention quite a departure from the current approach. The motivation behind sending a role descriptors intersection was to keep the contract for the headers generic enough to for example include other privileges in the future without having to change the underlying model, and also because that class is already established within the codebase.

> A more practical approach is to not use RemoteAccessRoleReference if the remote user is _system. We can have a different role reference type, e.g. StaticRoleReference

This was my thinking as well but that's the extra work I felt wasn't a big enough priority since it's more of an internal optimization that we can add later. In itself it's not that much work, but this PR is already quite complex and will be big once I've added all the requisite tests. Don't feel strongly though.

Reply:
Let's prioritize on getting things working end-to-end first. We can come back to this. For now, maybe just add this as an item under the "Refactoring oppoutunities" jira issue.